File: pa1/List.java

// Assignment Name: Programming Assignment 1 -- List ADT
// File: List.java 
// -----------------------------------------------------------------------------
// List.java is used to construct a List ADT using a sequence of finite integers. 
// It is used to store an array of strings from the provided input files and perform 
// manipulations on them.
// -----------------------------------------------------------------------------

public class List {

private int length;
private int index;
private Node back;
private Node front;
private Node cursor;

private class Node {
   Node prevNode;
   Node nextNode;
   int data;

   Node (int data)
   {
      prevNode=nextNode=null;

   }

   public String toString(){

         return String.valueOf(this.data);
      }


}

/*
The underlying data structure for the List ADT will be a doubly linked list. The List class should therefore contain
 a private inner Node class which itself contains fields for an int (the data), and two Node references (the previous
and next Nodes, respectively) which may be null. The Node class should also define an appropriate constructor
as well as a toString() method. 
*/ 

//contain private inner Node class, containing field for an int (data) and 2 nodes (pre and next)


      

      /*
      The List class should contain three private fields of type Node referring to the
front, back, and cursor elements, respectively. The List class should also contain private int fields storing the
length of the List and the index of the cursor element. 

in List() = When the cursor is undefined, an appropriate value the
index field is -1, since that is what is returned by index() in such a case.
*/
   

   List()
    {
      front = back = cursor = null;
      index = -1;
      length = 0;
   }

// Returns true if this List and L are the same integer
 // sequence. The cursor is ignored in both lists.
   boolean equals(List L)
    { if(L.length !=this.length)
      {
         return false;
      }

      Node N, M;
      List X;
      boolean eq;

      if (L instanceof List) {
         X = (List)L;
         M = this.front;
         N = X.front;
         eq = (this.length==X.length);

         while( eq!=false && N!=null ){
            eq = M.equals(N);
            M = M.nextNode;
            N = N.nextNode;
         }
 

      }
     return true;
   }
   // Access functions

   // Returns the number of elements in this List.
   int length()
   {
      return length;
   }

   // If cursor is defined, returns the index of the cursor element,
   // otherwise returns -1.
   int index() 
   {
    
         return index;
      
   }

   // Returns front element. Pre: length()>0
   int front() {
      if (length() > 0) {
         return front.data;
      }
      else
      {
         throw new RuntimeException ("List Error");
      }

   }

   // Returns back element. Pre: length()>0

   int back() {
      if (length() > 0) 
      {
         return back.data;
      }
      else
      {
         throw new RuntimeException ("List Error: back() called on an empty List");
      }
   }


   // Returns cursor element. Pre: length()>0, index()>=0

   int get()
   {
      if (length()>0 && index() >= 0){
         return cursor.data;
   }
      else{
         throw new RuntimeException(
            "List Error: get() called on empty List");
      }
      
   }

// Manipulation procedures

  

   // Resets this List to its original empty state.
   void clear() 
   {
      front = back = cursor = null;
      length = 0;
      index = -1;
   } 

   // If List is non-empty, places the cursor under the front element,
   // otherwise does nothing.
   void moveFront() 
   {
      if (this.length() > 0) 
      {
         cursor = front;
         index = 0;
      }
   }

   // If List is non-empty, places the cursor under the back element,
   // otherwise does nothing.
   void moveBack() {
      if (this.length() > 0) 
      {
         cursor = back;
         index = length()-1;
      }
   }

   // If cursor is defined and not at front, moves cursor one step toward
   // front of this List, if cursor is defined and at front, cursor becomes
   // undefined, if cursor is undefined does nothing.
   void movePrev() 
   {
      if (cursor != null)
       {
         cursor = cursor.prevNode;
         index--;
       } 
   } 

   // If cursor is defined and not at back, moves cursor one step toward
   // back of this List, if cursor is defined and at back, cursor becomes
   // undefined, if cursor is undefined does nothing.
   void moveNext() {

      if (cursor !=null) 
      {
         cursor = cursor.nextNode;
         index++;
      }

      if (cursor==null)
      {
         index = -1;
      }
   }

   // Insert new element into this List. If List is non-empty,
   // insertion takes place before front element.
void prepend(int data)
{  
   Node nNode=new Node(data);
   nNode.nextNode=front;

   if (front != null) 
   { 
      front.prevNode=nNode;
   }

   front=nNode;
   length++;

   if (cursor != null)
   {
      index++;
   }

   if(length==1)
   {
      back=nNode;
   }
   
}


   // Insert new element into this List. If List is non-empty,
   // insertion takes place after back element.
void append(int data) 
{  Node nNode=new Node(data);
   nNode.prevNode=back;

   if (length() > 0) 
   {
      back.nextNode = nNode;

   }
   back=nNode;
   length++;

   if(length ==1 )
   {
      front=nNode;
   }
}


   // Insert new element before cursor.
   // Pre: length()>0, index()>=0
void insertBefore(int data) {
   if (length() <= 0) {
      throw new RuntimeException(
         "List Error: insertBefore length called on an empty List");
      }
   else{
      Node nNode= new Node(data);
      nNode.prevNode = cursor.prevNode;
      nNode.nextNode=cursor;
      index++;
      length++;

      if(front==cursor)
      {
         front=nNode;

      }

      else 
      {
         cursor.prevNode.nextNode=nNode;
      }
   }
}

   // Inserts new element after cursor.
   // Pre: length()>0, index()>=0

void insertAfter(int data) {
   if (length() <= 0) {
      throw new RuntimeException(
         "List Error: insertAfter length called on empty List");
      }
   else if (index() < 0){
      throw new RuntimeException(
         "List Error: insertAfter index called on empty List");
   }

   Node nNode=new Node(data);
   nNode.nextNode =cursor.nextNode;
   nNode.prevNode=cursor;
   length++;


   if(cursor !=back)
   {
      cursor.nextNode.prevNode=nNode;
   }

   if(cursor==back)
   {
      back=nNode;
   }

   cursor.nextNode=nNode;
}
   // Deletes the front element. Pre: length()>0
   void deleteFront() {
   if (length()<=0) {
      throw new RuntimeException(
         "List Error: deleteFront length called on empty List");
   }
   else 
   {
      if (cursor == front)

      {
         cursor = null;
         index = -1;
      }

         front = front.nextNode;
        

         if (front !=null)
         {
             front.prevNode = null;
         }

         length--;
        if(cursor !=null)
        {
         index--;
        }

      }
   }

   // Deletes the back element. Pre: length()>0
   void deleteBack() 
   {

       if (length()<=0) 
   {

      throw new RuntimeException("List Error: deleteBack length called on empty List");
   }

      if (cursor==back)
      {
         cursor=null;
         index=-1;
         back=back.prevNode;
      }
      

      if (back !=null)
      {
         back.nextNode=null;
      }
      length--;
}
   

   // Deletes cursor element, making cursor undefined.
   // Pre: length()>0, index()>=0
   // Other methods
  void delete(){
   if (length()<=0) {
      throw new RuntimeException(
         "List Error: delete length called on empty List");
   }
   else if (index()<0) {
      throw new RuntimeException(
         "List Error: delete index called on empty List");
   } 
   else {
      if (cursor == front){
         deleteFront();
      }
      else if (cursor == back){
         deleteBack();
      }
      else if (this.length()>1)
       {
         cursor.prevNode.nextNode=cursor.nextNode;
         cursor.nextNode.prevNode=cursor.prevNode;
         index=-1;
         cursor=null;
         length--;
      }
   }
}


// Overrides Object's toString method. Returns a String
//  representation of this List consisting of a space
//  separated sequence of integers, with front on left.
public String toString(){
      StringBuffer sb = new StringBuffer();
      Node N = front;
      while(N!=null){
         sb.append(" ");
         sb.append(N.toString());
         N = N.nextNode;
      }
      return new String(sb);
   }
   


// Returns a new List representing the same integer sequence as this
 // List. The cursor in the new list is undefined, regardless of the
 // state of the cursor in this List. This List is unchanged.
List copy() {

      List copy = new List();
      Node N = front;

      while( N!=null ){
         copy.append(N.data);
         N = N.nextNode;
      }
      return copy;
      
   }


//not working
// List concat(List L)
// {
//     List nlist2=L.copy();

//    List nList=this.copy();

//    nList.back.nextNode=nList2.front;
//    nList2.front.prevNode=nList.back;
   
//    return nList;
// }

}

File: pa2/List.c

//PA2
//List.c
// -----------------------------------------------------------------------------
// List.c is used to construct a List ADT using a sequence of finite integers. 
// It is used to store an array of strings from the provided input files and perform 
// manipulations on them.
// -----------------------------------------------------------------------------
//List ADT implementation
//----------------------------------------------------------------------------

#include<stdio.h>
#include "List.h"
#include <stdlib.h>

typedef struct NodeObject{
    int data;
    struct NodeObject* prev;
    struct NodeObject* next;

}
NodeObject;
//private node typedef from java file
typedef  NodeObject* Node;
//private list object typdef in java file
typedef  struct ListObj{
    Node front;
    Node back;
    Node cursor;
    int length;
    int index;
} ListObj;
Node newNode(int data, Node prev, Node next)
{

    Node N = malloc(sizeof(NodeObject));
    N -> data= data;
    N -> prev= prev;
    N -> next=next;
    return(N);
}

// Constructors-Destructors ---------------------------------------------------
List newList(void){ //return a new list object
    List L;
    L = malloc(sizeof(ListObj));
    L->back = L->front = L->cursor= NULL;
    L->index = -1;
    L->length = 0;
    return (L);
}

// Access functions
int length(List L)// Returns the number of elements in this List.
{
    if (L==NULL)
    {
        printf ("List error, length cannot execute on a NULL List");
        exit(1);

    }
    return (L->length);
}
int isEmpty(List L){
    if( L==NULL ){
        printf("List Error: isEmpty() cannot be on NULL List reference");
        exit(1);
    }
    return(L->length==0);
}
int index(List L) // If cursor is defined, returns the index of the cursor element,
// otherwise returns -1.
{   if (L==NULL){
    printf ("List error, index() cannot execute on a NULL List");

}
    return(L->index);

}
int front(List L) // Returns front element. Pre: length()>0
{
    if (L == NULL)
    {
        printf("List error, front() cannot execute on a NULL List");
        exit(1);
    }

    return L->front->data;
}
int back(List L) // Returns back element. Pre: length()>0
{
    if (L == NULL)
    {
        printf("List error, back() cannot execute on a NULL List");
        exit(1);    }

    if (L->length <=0) //check if length is less than 0
    {
        printf("List error, back() cannot execute on a NULL List");
        exit(1);
    }

    return L->back->data;

}
int get(List L) // Returns cursor element. Pre: length()>0, index()>=0
{
    if (L == NULL)
    {
        printf("List error, get() cannot execute on a NULL List");
        exit(1);    }

    if (L->length <=0)
    {
        printf("List error, get() cannot execute on a NULL List");
        exit(1);
    }
    return L->cursor->data;

}
int equals(List A,List B) // Returns true if and only if this List and L are the same
// integer sequence. The states of the cursors in the two Lists
// are not used in determining equality.
{   int eq=0;
    Node First =NULL;
    Node Second=  NULL;
    if (A== NULL && B==NULL)
    {
        printf("List error, equals() cannot execute on a NULL List");
        exit(1);

    }
    eq= (A->length== B->length);
    First= A-> front; //iterators for both lists
    Second= B-> front;
    while(First && Second !=NULL) //checks if elements are equal
    {
        eq=(First->data==Second->data);
        First= First->next;
        Second=Second->next;

    }
    return eq;

}
// Manipulation procedures
void clear(List L) // Resets this List to its original empty state.
{
    if (L == NULL)
    {
        printf("List error, clear() cannot execute on a NULL List");
        exit(1);

    }
    else if(L!=NULL)
    {
        L->front=L->back=L->cursor=NULL;
        L->length=0;
        L->index=-1; //reset all variables to 0
    }
    while(length(L)>0)
    {
    	deleteFront(L);
    }
}
void moveFront(List L) // If List is non-empty, places the cursor under the front element,
// otherwise does nothing.
{
    if (L == NULL)
    {
        printf("List error, clear() cannot execute on a NULL List");
        exit(1);

    }
    if (!(isEmpty(L)))
    {
        L->cursor=L->front;
        L->index=0;
    }
}
void moveBack(List L) // If List is non-empty, places the cursor under the back element,
// otherwise does nothing.
{
    if (L == NULL)
    {
        printf("List error, moveBack() cannot execute on a NULL List");
        exit(1);

    }
    if (!(isEmpty(L)))
    {
        L->cursor=L->back;
        L->index=L->length-1;
    }
}
void movePrev(List L) // If cursor is defined and not at front, moves cursor one step toward
// front of this List, if cursor is defined and at front, cursor becomes
// undefined, if cursor is undefined does nothing.
{
    if (L == NULL)
    {
        printf("List error, movePrev() cannot execute on a NULL List");
        exit(1);
    }//move cursor back one node
    if(L->cursor==L->front)
    { L-> cursor=NULL;
        L->index=-1;
    }
    else if (L->cursor !=NULL)
    {
        L->cursor=L->cursor->prev;
        L->index--;
    }

}
void moveNext(List L) // If cursor is defined and not at back, moves cursor one step toward
// back of this List, if cursor is defined and at back, cursor becomes
// undefined, if cursor is undefined does nothing.
{  if (L == NULL)
{
    printf("List error, Movenext() cannot execute on a NULL List");
    exit(1);
}
    if (L->cursor== L->back)
    { L->cursor=NULL;
        L->index= -1;
    }
    if (L->cursor!= NULL)
    {
      L->cursor=L->cursor->next;
      L->index++;
    }
}
void prepend(List L, int data) // Insert new element into this List. If List is non-empty,
// insertion takes place before front element.
{
    if (L == NULL)
    {
        printf("List error, prepend() cannot execute on a NULL List");
        exit(1);
    }
    Node N = newNode(data, NULL, L->front);

    if(!isEmpty(L)) //point front and back to new node
    {
        L->front->prev=N;
        L->front=N;
        if (index(L)!=-1)
        {
            L ->index++;
        }

    }
    else if (length(L) == 0) //make N the front
    {
        L->back= L->front=N;
    }
     N=NULL;
    L->length++;

}

void append(List L, int data) // Insert new element into this List. If List is non-empty,
// insertion takes place after back element.
{
  if (L == NULL)
  {
      printf("List error, prepend() cannot execute on a NULL List");
      exit(1);
  }
  Node N = newNode(data, L->back, NULL);

  if(!isEmpty(L))
  {
      L->back->next=N;
      L->back=N;
   }
   else if (length(L) == 0)
   {
       L->front= L->back=N;
   }

   N=NULL;
   L->length++;
}

void insertBefore(List L, int data) // Insert new element before cursor.
// Pre: length()>0, index()>=0
{
    if(L==NULL)
    {
        printf("List error, insertBefore() cannot execute on a NULL List");
        exit(1);

    }
    if (length(L)==0 || L->cursor == L->front)
    {
        prepend(L,data);
        return;

    }
    else
    {
        Node old= newNode(data, L->cursor->prev, L->cursor);
        L->cursor->prev->next=old;
        L->cursor->prev=old;
    }
    L->length++;
    L->index++;

}
void insertAfter(List L, int data) // Inserts new element after cursor.
// Pre: length()>0, index()>=0
{
    if(L==NULL)
    {
        printf("List error, insertAfter() cannot execute on a NULL List");
        exit(1);

    }
    L->length++;
    Node post=newNode(data, L->cursor, L->cursor->next);
    if (L->cursor !=L->back)
    {
        L->cursor->next->prev=post;
    }
    L->cursor->next=post;
    if(L->cursor==L->back)
    {
        L->back=post;
    }
}
void deleteFront(List L) // Deletes the front element. Pre: length()>0
{
    if(L==NULL)
    {
        printf("List error, deletefront() cannot execute on a NULL List");
        exit(1);

    }
    if (L->length<=1)
    {
        L->front=L->back=NULL; //delete front

    }
    else
    {
        L->front=L->front->next;
        L->front->prev=NULL;
    }
    L->length--;
}
void deleteBack(List L) // Deletes the back element. Pre: length()>0
{   if(L==NULL)
{
    printf("List error, deleteBack() cannot execute on a NULL List");
    exit(1);

}    if(L->length>1)
{
    L->back=L->back->prev; //switch back with previous
    L->back->next=NULL;
}
else if (L->length==1)
{
    L->front=L->back=NULL;
}
    L->length--;
}

void delete(List L)
{
    if(L==NULL)
    {
        printf("List error, deleteBack() cannot execute on a NULL List");
        exit(1);
    }
    if (L->length>1)
    {
        L->cursor=L->cursor->next;

    }
    else if (L->length<1)
    {
        L->front=L->back=NULL; //delete back element
    }
    L->length--;

}
List copyList(List L){
    if(L==NULL)
    {
        printf("List error, copylist() cannot execute on a NULL List");
        exit(1);
    }
    List newlist= newList(); //create a list for copy destination
    if(length(L)==0)
        return(newlist);
    Node N=NULL;
    N=L->front;

    while(N != NULL) {
        append(newlist, N->data);
        N = N->next;
    }
    return(newlist);
}
void freeList(List* pL)
    {
        if(pL!=NULL && *pL!=NULL) {
            while( !isEmpty(*pL) ) {
                deleteFront(*pL);

            }
            free(*pL);
            *pL = NULL;
        }
}
void printList (FILE* out, List L)
{
    Node N=NULL;
    if(L==NULL)
    {
        printf("List error, copylist() cannot execute on a NULL List");
        exit(1);
    }
    if(out==NULL)
    {
        printf("List error, copylist() cannot execute on a NULL List");
        exit(1);
    }
    N=L->front;


    while(N != NULL)
    {
        fprintf(out, "%d", N->data);
        if (N->next != NULL) {
          fprintf(out, " ");
        }
        N= N->next;
    }
    // printf("\n");
}
